File: backend/src/main/java/cn/etarch/mao/auth/service/AuthService.java
package cn.etarch.mao.auth.service;

import cn.etarch.mao.auth.controller.AuthController.*;
import cn.etarch.mao.common.exception.BusinessException;
import cn.etarch.mao.common.result.ErrorCode;
import cn.etarch.mao.user.entity.User;
import cn.etarch.mao.user.mapper.UserMapper;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserMapper userMapper;
    private final JwtService jwtService;
    private final PasswordEncoder passwordEncoder;
    private final LdapAuthService ldapAuthService;

    public LoginVO login(String username, String password) {
        User user = userMapper.selectOne(
                new QueryWrapper<User>().eq("username", username));

        // 本地密码验证
        if (user != null && user.getPasswordHash() != null
                && passwordEncoder.matches(password, user.getPasswordHash())) {
            return buildLoginResult(user);
        }

        // LDAP 回退验证
        if (ldapAuthService != null && ldapAuthService.isConfigured()) {
            try {
                return ldapAuthService.authenticate(username, password);
            } catch (Exception ignored) {
            }
        }

        throw new BusinessException(ErrorCode.LOGIN_FAILED);
    }

    private LoginVO buildLoginResult(User user) {
        if (user.getStatus() != null && user.getStatus() == 0) {
            throw new BusinessException(ErrorCode.ACCOUNT_DISABLED);
        }

        user.setLastLoginAt(LocalDateTime.now());
        userMapper.updateById(user);

        String accessToken = jwtService.generateToken(user.getId(), user.getUsername());
        String refreshToken = jwtService.generateRefreshToken(user.getId(), user.getUsername());

        LoginVO vo = new LoginVO();
        vo.setAccessToken(accessToken);
        vo.setRefreshToken(refreshToken);
        vo.setExpiresIn(86400L);

        UserInfoVO userInfo = new UserInfoVO();
        userInfo.setId(user.getId());
        userInfo.setUsername(user.getUsername());
        userInfo.setDisplayName(user.getDisplayName());
        userInfo.setEmail(user.getEmail());
        userInfo.setAvatarUrl(user.getAvatarUrl());
        vo.setUser(userInfo);

        return vo;
    }

    public LoginVO refreshToken(String refreshToken) {
        if (!jwtService.validateToken(refreshToken)) {
            throw new BusinessException(ErrorCode.TOKEN_EXPIRED);
        }

        Long userId = jwtService.getUserIdFromToken(refreshToken);
        String username = jwtService.getUsernameFromToken(refreshToken);

        User user = userMapper.selectById(userId);
        if (user == null || (user.getStatus() != null && user.getStatus() == 0)) {
            throw new BusinessException(ErrorCode.ACCOUNT_DISABLED);
        }

        String newAccessToken = jwtService.generateToken(userId, username);
        String newRefreshToken = jwtService.generateRefreshToken(userId, username);

        LoginVO vo = new LoginVO();
        vo.setAccessToken(newAccessToken);
        vo.setRefreshToken(newRefreshToken);
        vo.setExpiresIn(86400L);

        UserInfoVO userInfo = new UserInfoVO();
        userInfo.setId(user.getId());
        userInfo.setUsername(user.getUsername());
        userInfo.setDisplayName(user.getDisplayName());
        userInfo.setEmail(user.getEmail());
        userInfo.setAvatarUrl(user.getAvatarUrl());
        vo.setUser(userInfo);

        return vo;
    }

    public void logout() {
        // Stateless JWT logout is handled client-side by discarding the token.
    }
}

File: backend-ts/src/auth/auth.service.ts
import { BusinessException } from '../common/business-exception.js';
import { ErrorCode } from '../common/error-code.js';
import { JwtService } from '../crypto/jwt.service.js';
import { UserService } from '../user/user.service.js';
import type { LoginVO, PasswordHasher, User, UserInfoVO, UserRepository } from '../user/types.js';
import type { PermissionService } from '../permission/permission.service.js';
import type { LdapAuthService } from './ldap-auth.service.js';

export class AuthService {
  constructor(
    private readonly userRepo: UserRepository,
    private readonly jwtService: JwtService,
    private readonly passwordHasher: PasswordHasher,
    private readonly ldapAuthService: LdapAuthService | null,
    private readonly permissionService?: PermissionService,
  ) {}

  async login(username: string, password: string): Promise<LoginVO> {
    const user = await this.userRepo.findByUsername(username);
    if (user && user.passwordHash && (await this.passwordHasher.matches(password, user.passwordHash))) {
      return this.buildLoginResult(user);
    }
    if (this.ldapAuthService?.isConfigured()) {
      try {
        return await this.ldapAuthService.authenticate(username, password);
      } catch {
        // fall through
      }
    }
    throw new BusinessException(ErrorCode.LOGIN_FAILED);
  }

  async buildLoginResult(user: User): Promise<LoginVO> {
    if (user.status != null && user.status === 0) {
      throw new BusinessException(ErrorCode.ACCOUNT_DISABLED);
    }
    user.lastLoginAt = formatNow();
    await this.userRepo.updateById(user);
    return this.toLoginVO(user);
  }

  async refreshToken(refreshToken: string): Promise<LoginVO> {
    if (!this.jwtService.validateToken(refreshToken)) {
      throw new BusinessException(ErrorCode.TOKEN_EXPIRED);
    }
    const userId = this.jwtService.getUserIdFromToken(refreshToken);
    const username = this.jwtService.getUsernameFromToken(refreshToken);
    const user = await this.userRepo.findById(userId);
    if (!user || (user.status != null && user.status === 0)) {
      throw new BusinessException(ErrorCode.ACCOUNT_DISABLED);
    }
    void username;
    return this.toLoginVO(user);
  }

  logout(): void {
    // Stateless JWT logout is handled client-side by discarding the token.
  }

  async toUserInfoVO(user: User, withPerms = false): Promise<UserInfoVO> {
    const vo: UserInfoVO = {
      id: user.id,
      username: user.username,
      displayName: user.displayName,
      email: user.email,
      avatarUrl: user.avatarUrl,
      authSource: UserService.resolveAuthSource(user),
    };
    if (withPerms && this.permissionService && user.id != null) {
      const codes = await this.permissionService.getUserPermissionCodes(user.id);
      vo.permissions = [...new Set(codes)];
      vo.isAdmin = await this.permissionService.isAdmin(user.id);
    }
    return vo;
  }

  private async toLoginVO(user: User): Promise<LoginVO> {
    const accessToken = this.jwtService.generateToken(user.id!, user.username);
    const refreshToken = this.jwtService.generateRefreshToken(user.id!, user.username);
    return {
      accessToken,
      refreshToken,
      expiresIn: 86400,
      user: await this.toUserInfoVO(user),
    };
  }
}

export function formatNow(): string {
  const d = new Date();
  const pad = (n: number) => String(n).padStart(2, '0');
  const tz = new Date(d.toLocaleString('en-US', { timeZone: 'Asia/Shanghai' }));
  return `${tz.getFullYear()}-${pad(tz.getMonth() + 1)}-${pad(tz.getDate())} ${pad(tz.getHours())}:${pad(tz.getMinutes())}:${pad(tz.getSeconds())}`;
}
